File: frontend/src/lib/api/schemas/approvals.ts
import { z } from "zod";
import { ApprovalKind, ApprovalStatus } from "@/types/enums";

const approvalKindSchema = z.enum([
  ApprovalKind.PUBLISH,
  ApprovalKind.SCHEDULE,
  ApprovalKind.VOICE_CLONING,
  ApprovalKind.ASSET_DELETION,
]);

const approvalStatusSchema = z.enum([
  ApprovalStatus.PENDING,
  ApprovalStatus.APPROVED,
  ApprovalStatus.REJECTED,
]);

export const approvalReadSchema = z.object({
  id: z.string().uuid(),
  kind: approvalKindSchema,
  status: approvalStatusSchema,
  payload: z.record(z.string(), z.unknown()),
  workflow_id: z.string().nullable(),
  requested_by_user_id: z.string().uuid().nullable(),
  decided_by_user_id: z.string().uuid().nullable(),
  decided_at: z.string().nullable(),
  note: z.string().nullable(),
});
export type ApprovalRead = z.infer<typeof approvalReadSchema>;

export type ApprovalRequestRequest = {
  kind: ApprovalKind;
  payload?: Record<string, unknown>;
  workflow_id?: string | null;
};

export type ApprovalDecisionRequest = {
  status: typeof ApprovalStatus.APPROVED | typeof ApprovalStatus.REJECTED;
  note?: string | null;
};

File: backend/src/ytforge/interfaces/api/schemas/approvals.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from ytforge.domain.enums import ApprovalKind, ApprovalStatus


class ApprovalRequestRequest(BaseModel):
    kind: ApprovalKind
    payload: dict[str, Any] = Field(default_factory=dict)
    workflow_id: str | None = None


class ApprovalDecisionRequest(BaseModel):
    status: ApprovalStatus
    note: str | None = None


class ApprovalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    kind: ApprovalKind
    status: ApprovalStatus
    payload: dict[str, Any]
    workflow_id: str | None
    requested_by_user_id: uuid.UUID | None
    decided_by_user_id: uuid.UUID | None
    decided_at: datetime | None
    note: str | None
